reinforced_lib/utils/experience_replay.py

```diff
@@ -0,0 +1,202 @@
+    """
+
+    states: Array
+    actions: Array
+    rewards: Array
+    terminals: Array
+    next_states: Array
+    size: jnp.int32
+    ptr: jnp.int32
+
+
+def experience_replay(
+        buffer_size: jnp.int32,
+        batch_size: jnp.int32,
+        obs_space_shape: Shape,
+        act_space_shape: Shape
+) -> ExperienceReplay:
+    """
+    Experience replay buffer used for off-policy learning. Improves the stability of the learning process
+    by reducing the correlation between the samples and enables an agent to learn from past experiences.
+
+    Parameters
+    ----------
+    buffer_size : int
+        Maximum size of the replay buffer.
+    batch_size : int
+        Size of the batch to be sampled from the replay buffer.
+    obs_space_shape : Shape
+        Shape of the observation space.
+    act_space_shape : Shape
+        Shape of the action space.
+
+    Returns
+    -------
+    out : ExperienceReplay
+        Container for experience replay buffer functions.
+    """
+
+    def init() -> ReplayBuffer:
+        """
+        Initializes the replay buffer with empty arrays.
+
+        Returns
+        -------
+        buffer : ReplayBuffer
+            Dataclass containing the replay buffer values.
+        """
+
+        return ReplayBuffer(
+            states=jnp.empty((buffer_size, *obs_space_shape)),
+            actions=jnp.empty((buffer_size, *act_space_shape)),
+            rewards=jnp.empty(buffer_size),
+            terminals=jnp.empty(buffer_size, dtype=jnp.bool_),
+            next_states=jnp.empty((buffer_size, *obs_space_shape)),
+            size=0,
+            ptr=0
+        )
+
+    def append(
+            buffer: ReplayBuffer,
+            state: Numeric,
+            action: Numeric,
+            reward: Scalar,
+            terminal: jnp.bool_,
+            next_state: Numeric
+    ) -> ReplayBuffer:
+        """
+        Appends a new values to the replay buffer.
+
+        Parameters
+        ----------
+        buffer : ReplayBuffer
+            Dataclass containing the replay buffer values.
+        state : array_like
+            State of the environment.
+        action : array_like
+            Action taken by the agent.
+        reward : float
+            Reward received by the agent.
+        terminal : bool
+            Flag indicating if the episode has terminated.
+        next_state : array_like
+            Next state of the environment.
+
+        Returns
+        -------
+        buffer : ReplayBuffer
+            Updated replay buffer.
+        """
+
+        return ReplayBuffer(
+            states=buffer.states.at[buffer.ptr].set(state),
+            actions=buffer.actions.at[buffer.ptr].set(action),
+            rewards=buffer.rewards.at[buffer.ptr].set(reward),
+            terminals=buffer.terminals.at[buffer.ptr].set(terminal),
+            next_states=buffer.next_states.at[buffer.ptr].set(next_state),
+            size=buffer.size + 1,
```

Comment:
Why not replace it with: `size = min(buffer_size, buffer.size + 1)`, than in line 169 could be simplified to just `choice_size = buffer.size`

Reply:
I agree, thanks for the advice!